### PR TITLE
feat: add alert rules for mso

### PR DESCRIPTION
**Pull request by slashpai**
This commit adds initial rule file for monitoring-stack-operator

Signed-off-by: Jayapriya Pai janantha@redhat.com

**Description:**

Add alert rules for controller, create service and servicemonitor

### CODE REVIEW

File: deploy/samples/monitoring-stack.yaml [hunk 1]
```diff
@@ -7,4 +7,4 @@ spec:
   retention: 1d
   resourceSelector:
     matchLabels:
-      system: foo
+      app.kubernetes.io/part-of: monitoring-stack-operator
```
Comment (slashpai) [inline deploy/samples/monitoring-stack.yaml:10]: I updated this to `app.kubernetes.io/part-of: monitoring-stack-operator` since resourceSelector label need to match the object label while creating prometheus cr's, I either needed to add `system:foo` to cr's or just update the label here. Looking for better way to handle if this is not the right way

**Comment by slashpai** [comment]
@fpetkovski @sthaha PTAL

File: deploy/samples/monitoring-stack.yaml (hunk 1 above)
Comment (fpetkovski) [reply deploy/samples/monitoring-stack.yaml:10]: Hi JP, I wonder if we the stacks should collect these service monitors. I would expect them to end up in some other prometheus since they're related to the operator itself. So I don't think we should worry too much about matching the selectors of the stack resource and the service monitor(s) for the operator.

Comment (slashpai) [reply deploy/samples/monitoring-stack.yaml:10]: sure I added that since I was trying to test alert I will remove that. We want to keep only rules file?

Comment (fpetkovski) [reply deploy/samples/monitoring-stack.yaml:10]: I think it's fine to keep the ServiceMonitor as well. Let's just make sure it selects the component label from the pods, not just the `app.kubernetes.io/name`:
https://github.com/rhobs/monitoring-stack-operator/blob/main/deploy/operator/monitoring-stack-operator-deployment.yaml#L18-L20

Comment (slashpai) [reply deploy/samples/monitoring-stack.yaml:10]: updated, I generated bundle locally it works fine

I looked at https://docs.openshift.com/container-platform/4.9/operators/understanding/olm-packaging-format.html#olm-bundle-format_olm-packaging-format servicemonitor and prometheusrule are supported for bundle

So a question is since the service, servicemonitor and prometheusrule will be only used when prometheus resource is created deploying them by default is ok?

Comment (fpetkovski) [reply deploy/samples/monitoring-stack.yaml:10]: I think deploying them by default is okay, we just have to make sure that they don't get picked up by some prometheus deployed by a monitoring stack. 

We should probably document how to collect these monitoring resources in a Prometheus instance, I think SRE will definitely be interested in that.

Comment (slashpai) [reply deploy/samples/monitoring-stack.yaml:10]: Without a matching resourceSelector label it doesn't get picked up by default, cant we use that logic? I meant while creating monitoringstack cr we can specify the label we should use as in this sample-monitoring stack file?

File: deploy/olm/kustomization.yaml [hunk 2]
```diff
@@ -10,4 +10,4 @@ kind: Kustomization
 images:
 - name: monitoring-stack-operator
   newName: monitoring-stack-operator
-  newTag: 0.0.2
+  newTag: 0.0.6
```
Comment (sthaha) [inline deploy/olm/kustomization.yaml:13]: Unrelated change?

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    prometheus: k8s
+    role: alert-rules
+  name: monitoring-stack-operator-prometheus-rules
+spec:
+  groups:
+  - name: monitoring-stack-operator.rules
+    rules:
+    - alert: MonitoringStackOperatorReconcileErrors
+      annotations:
+        description: MonitoringStackOperator failing to reconcile controller {{ $labels.controller
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:17]: How about `Monitoring Stack Operator controller - {{$labels.controller}} fails to reconcile`

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+              }}
+        summary: MonitoringStack Operator failing to reconcile controller {{ $labels.controller
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:19]: Same - `Monitoring Stack Operator controller - {{$labels.controller}} fails to reconcile`

File: deploy/operator/monitoring-stack-operator-service-monitor.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: monitoring-stack-operator
+  labels:
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-service-monitor.yaml:5]: I wonder if we can make use of `commonLabels` in `kustomize.yaml` ?

File: deploy/samples/monitoring-stack.yaml (hunk 1 above)
Comment (sthaha) [reply deploy/samples/monitoring-stack.yaml:10]: Since the goal of this PR is to generate alerts if **MSO** fails to reconcile, should we worry about the stacks we deploy? Not only that, these rules don't apply to stack we deploy. Hence ... we shouldn't need to modify the sample stack - cr

File: deploy/olm/kustomization.yaml (hunk 2 above)
Comment (slashpai) [reply deploy/olm/kustomization.yaml:13]: ya i can remove that, got update when I ran `make bundle`

File: deploy/operator/monitoring-stack-operator-service-monitor.yaml (hunk 5 above)
Comment (slashpai) [reply deploy/operator/monitoring-stack-operator-service-monitor.yaml:5]: ya we can, I thought to have it in a separate PR since I need to update other files also  unrelated to this  change

File: deploy/samples/monitoring-stack.yaml (hunk 1 above)
Comment (slashpai) [reply deploy/samples/monitoring-stack.yaml:10]: The prometheus resources servicemonitor, prometheusrule were not picking by operator unless the resourceSelector matching label is there. I had updated to test this change but I can remove.

Comment (slashpai) [reply deploy/samples/monitoring-stack.yaml]: I reverted it to same as before but to confirm we can use this logic to make sure it doesn't picked by monitoring-stack-operator by default unless a matchingLabel found?

File: deploy/operator/monitoring-stack-operator-service-monitor.yaml (hunk 5 above)
Comment (fpetkovski) [reply deploy/operator/monitoring-stack-operator-service-monitor.yaml:5]: The problem with commonLabels is that it sets selectors as well. So we have to make sure we don't have any values that change (e.g. the version) in those labels.

Comment (slashpai) [reply deploy/operator/monitoring-stack-operator-service-monitor.yaml:5]: ya we can keep component, name and part-of?

Comment (fpetkovski) [reply deploy/operator/monitoring-stack-operator-service-monitor.yaml:5]: I think we can only have `part-of`, no? Otherwise every component will have the same name and same component.

File: deploy/samples/monitoring-stack.yaml (hunk 1 above)
Comment (slashpai) [reply deploy/samples/monitoring-stack.yaml]: @fpetkovski wdyt?

**Comment by fpetkovski** [comment]
This looks good so far. Do we want to monitor additional metrics, e.g. queue depth?

**Comment by slashpai** [comment]
> This looks good so far. Do we want to monitor additional metrics, e.g. queue depth?

Ya I will add other rules as well, created initial PR with only one just to get an idea. Also since we are not using jsonnet how do we want to include other alerts like monitoring prometheus-operator?

**Comment by fpetkovski** [comment]
Yes, that's a good point. We should include PO alerts as well.

**Comment by slashpai** [comment]
@fpetkovski I added couple of alerts as well, PTAL. I can create separate PR for Prometheus Operator

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+spec:
+  groups:
+  - name: monitoring-stack-operator.rules
+    rules:
+    - alert: MonitoringStackOperatorReconcileErrors
+      annotations:
+        description: Monitoring Stack Operator controller - {{ $labels.controller }} fails to reconcile. Inspect the
+          monitoring-stack-operator log for potential root causes.
+        summary: Monitoring Stack Operator controller - {{ $labels.controller }} fails to reconcile
+      expr: increase(controller_runtime_reconcile_errors_total{job="monitoring-stack-operator"}[15m])
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:20]: nit: could you consider making use of `|` or `>` to make this more readable 

e.g.
```yaml
description: >- # or even |-
  Monitoring Stack Operator controller - {{ $labels.controller }} fails to reconcile. 
  Inspect the monitoring-stack-operator log for potential root causes.
expr: |
  increase(controller_runtime_reconcile_errors_total{job="monitoring-stack-operator"[15m]} > 0

 
```

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        > 0
+      for: 15m
+      labels:
+        severity: warning
+    - alert: MonitoringStackOperatorReconcilation takes longer than 10 minutes
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:25]: `Monitoring Stack Operator reconciliation`  to be consistent with the rest of the alerts

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      annotations:
+        description: MonitoringStackOperator reconcilation takes longer than 10 minutes for the controller - {{ $labels.controller }}.
+          Inspect the monitoring-stack-operator log for potential root causes.
+        summary: MonitoringStackOperator reconcilation takes longer for controller - {{ $labels.controller }}
+      expr: rate(controller_runtime_reconcile_time_seconds_sum{job="monitoring-stack-operator"}[5m]) /
+            rate(controller_runtime_reconcile_time_seconds_count{job="monitoring-stack-operator"}[5m]) > 600
+      for: 10m
+      labels:
+        severity: warning
+    - alert: MonitoringStackController backlog is not being drained
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:35]: Monitoring Stack Operator

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      annotations:
+        description: The backlog for {{ $labels.name }} is not getting drained. It is an indication of reconciliation loop of controller got stuck.
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:37]: > It is an indication of reconciliation loop of controller got stuck.

The backlog of Monitoring Stack Operator - controller {{ $labels.name }} is not getting drained; an indication that reconcile loop may be stuck.

**Review by sthaha — COMMENTED** [review]
Nice patch @slashpai ! 
Looks good overall. Some minor corrections

**Comment by slashpai** [comment]
> Nice patch @slashpai ! Looks good overall. Some minor corrections

Thanks Sunil. I updated with changes, PTAL

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        increase(controller_runtime_reconcile_errors_total{job="monitoring-stack-operator"}[15m]) > 0
+      for: 15m
+      labels:
+        severity: warning
+    - alert: Monitoring Stack Operator reconcilation takes longer than 10 minutes
+      annotations:
+        description: |
+          Monitoring Stack Operator reconcilation takes longer than 10 minutes for the controller - {{ $labels.controller }}.
+          Inspect the monitoring-stack-operator log for potential root causes.
+        summary: Monitoring Stack Operator reconcilation takes longer for controller - {{ $labels.controller }}
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:31]: > takes a long time to reconcile 
or to be consisten ... 

`Monitoring Stack Operator controller - {{ $labels.controller }} takes too long to reconcile`

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        > 600
+      for: 10m
+      labels:
+        severity: warning
+    - alert: Monitoring Stack Operator backlog is not being drained
+      annotations:
+        description: |
+          The backlog of Monitoring Stack Operator - {{ $labels.name }} is not getting drained; an indication that reconcile loop may be stuck
+          Inspect the monitoring-stack-operator log for potential root causes.
+        summary: Monitoring Stack Operator - {{ $labels.name }} backlog is not being drained.
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:44]: `Monitoring Stack Operator controller - {{ $labels.name }}` - right? Isn't `labels.name` the controller name?

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        rate(controller_runtime_reconcile_time_seconds_sum{job="monitoring-stack-operator"}[5m]) /
+        rate(controller_runtime_reconcile_time_seconds_count{job="monitoring-stack-operator"}[5m])
+        > 600
+      for: 10m
+      labels:
+        severity: warning
+    - alert: Monitoring Stack Operator backlog is not being drained
+      annotations:
+        description: |
+          The backlog of Monitoring Stack Operator - {{ $labels.name }} is not getting drained; an indication that reconcile loop may be stuck
```
Comment (sthaha) [inline deploy/operator/monitoring-stack-operator-rules.yaml:42]: nit: We can be consistent in our alert messages .. e.g. 

Monitoring Stack Operator controller - {{ $labels.name }}  backlog  is no getting drained; an indication that reconcile loop may be stuck.

File: deploy/operator/monitoring-stack-operator-rules.yaml (hunk 11 above)
Comment (slashpai) [reply deploy/operator/monitoring-stack-operator-rules.yaml:44]: yes its controller name

**Review by sthaha — APPROVED** [review]
nice!

**Comment by slashpai** [comment]
@fpetkovski are we good to merge? or anymore change is required :)

File: deploy/operator/monitoring-stack-operator-rules.yaml [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        severity: warning
+    - alert: Monitoring Stack Operator controller reconcilation takes longer than 10 minutes
+      annotations:
+        description: |
+          Monitoring Stack Operator controller reconcilation takes longer than 10 minutes for the controller - {{ $labels.controller }}.
+          Inspect the monitoring-stack-operator log for potential root causes.
+        summary: Monitoring Stack Operator controller - {{ $labels.controller }} reconcilation takes too long to reconcile
+      expr: |
+        rate(controller_runtime_reconcile_time_seconds_sum{job="monitoring-stack-operator"}[5m]) /
+        rate(controller_runtime_reconcile_time_seconds_count{job="monitoring-stack-operator"}[5m])
```
Comment (fpetkovski) [inline deploy/operator/monitoring-stack-operator-rules.yaml:34]: What is a rate of a count in this case?

Comment (fpetkovski) [reply deploy/operator/monitoring-stack-operator-rules.yaml:34]: Ah ok so this is the average loop time, makes sense :+1:

**Review by fpetkovski — APPROVED** [review]
